pkg/block/s3/adapter.go

```diff
@@ -258,13 +271,14 @@ func (a *Adapter) Put(ctx context.Context, obj block.ObjectPointer, sizeBytes in
 		a.registerCaptureServerMiddleware(),
 	)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	etag := aws.ToString(resp.ETag)
 	if etag == "" {
-		return ErrMissingETag
+		return nil, ErrMissingETag
 	}
-	return nil
+	mtime := getServerTimeFromResponseMetadata(resp.ResultMetadata)
```

Comment:
AFAICT this is still incorrect 😭 .  This code ends up taking the time in the server `Date` header.  But I could not find anywhere where AWS say that this is the actual "Last-Modified" time of the object on S3.  Indeed, it might not be.

- The claim that "but it will be later" does not settle the matter.  Right now a user has issues with If-Unmodified-Since, where obviously reporting a later time is better.  But S3 get-object _also_ supports If-Modified-Since, which swings in exactly the opposite direction.
- The claim that "time on the AWS gateway will be close to time on the S3 backend" is much stronger.  But still obviously not enough - AWS do _not_ consider the 2 to be equivalent, see e.g. copy-object which returns a LastModified XML element in addition to the standard Date header.

I would much prefer to have an extra head-object call here when needed; obviously whether to do that would have to depend on some PutOpts so we don't do it and drop.

Reply:
This is a compromise we took under consideration. By definition lakeFS mtime will always be explicitly >= storage mtime. A good example for that is modifying metadata in lakeFS where we modify the mtime in lakeFS but storage mtime remains unmodified. We do not expect equality.